src/github.com/stellar/horizon/resource/account.go

```diff
@@ -71,3 +72,16 @@ func (this *Account) Populate(
 	this.Links.Data.PopulateTemplated()
 	return
 }
+
+// GetData returns decoded value for a given key. If the key does
+// not exist, empty slice will be returned.
+func (this *Account) GetData(key string) []byte {
+	data, exists := this.Data[key]
+	if !exists {
+		return nil
+	}
+
+	// We assume that horizon always returns valid base64 encoded string
+	decoded, _ := base64.StdEncoding.DecodeString(data)
```

Comment:
It's quite late here, so I will keep it short and we can discuss it tomorrow during standup.
> We can't validate the data value ahead of time because it is being loaded from stellar-core's database. It is an external system to horizon and can contain anything as far as horizon is concerned. stellar-core could introduce a bug that corrupts that field; we have no way of ensuring that data is and will always be valid.

We can at least validate certain things like encoding (ex. data value) just before displaying it to the user but to check if the value is really correct we can do what I described below.

> For example, we have no choice but to trust the balance fields from the stellar-core database. There's no easy way that I know of the validate a balance other than contacting an independent stellar-core instance.

I think it can be solved by hardcoding the first few ledger hashes and account balances for a given network (because master account secret is known by everyone) and then checking signatures on transactions, hashes of new ledgers and validating operations (tracking flows between accounts, trustlines etc). It duplicates a lot of things stellar-core does but I'm just arguing it's possible without another stellar-core. (Of course, with access to stellar-core db the attacker would be able to do nasty things but things that would only affect the account he/she owns - has secret keys).

The key question here is: do we want to do all of this (making horizon another layer of defence) and what degree of trust users should have for horizon? I hope we will find answers tomorrow.